### PR TITLE
Adding Customisation - insertQuery for postgres/mysql 

### PR DESCRIPTION
**Pull request by LazyDBA247-Anyvision**
Added the ability to write your own insertQuery to support the UPSERT syntax most of the RDBMS in now day already support.

### CODE REVIEW

**Comment by kmlebedev** [comment]
@LazyDBA247-Anyvision 
It seems like a bad idea to add it to the config `insertQuery`
It might be better to check the RDBMS version and choose the insert type automatically ?

File: weed/command/scaffold.go [hunk 1]
```diff
@@ -102,6 +102,10 @@ enabled = false
 dir = "./filerrdb"					# directory to store rocksdb files
 
 [mysql]  # or memsql, tidb
+# Empty insertQuery will use the default insert query
+# example insertQuery can be for UPSERT syntax:
+# insertQuery = """INSERT INTO ` + "`%s`" + ` (dirhash,name,directory,meta) VALUES(?,?,?,?) ON DUPLICATE KEY UPDATE meta = VALUES(meta)"""
```
Comment (chrislusf) [inline weed/command/scaffold.go:107]: The example seems wrong. ``` ` + "`%s`" + ` ``` should not need the plus operator for the string concatenation.

Comment (LazyDBA247-Anyvision) [reply weed/command/scaffold.go:107]: that is because the scaffold uses \` and MySQL uses \` to quote the table name, but go uses it as string literal...

**Comment by LazyDBA247-Anyvision** [comment]
@kmlebedev postgres / mysql driver fits many databases and for each, you may have some syntax differences.
until today the old way worked but caused 2 round trips from filer to DB and 2 table seeks. I am offering customization that needs to be verified and maybe even written by the owner/deployer.

**Comment by chrislusf** [comment]
> @LazyDBA247-Anyvision
> It seems like a bad idea to add it to the config `insertQuery`
> It might be better to check the RDBMS version and choose the insert type automatically ?

There are many mysql/postgres variations.

File: weed/command/scaffold.go (hunk 1 above)
Comment (LazyDBA247-Anyvision) [reply weed/command/scaffold.go:107]: @chrislusf  already exists and working in:
https://github.com/chrislusf/seaweedfs/blob/96c62bd34d4d2c99dea3164a9c062abaddb4eaab/weed/command/scaffold.go#L135

**Comment by LazyDBA247-Anyvision** [comment]
@chrislusf that's exactly why I added it as an optional and the possibility to update the syntax to fit your specific variant.

**Comment by kmlebedev** [comment]
@LazyDBA247-Anyvision I understand, but this raises the entry threshold for configuring inserts.
I believe it could have been made easier!

**Comment by LazyDBA247-Anyvision** [comment]
@kmlebedev fully backward compatible and without breaking new deployment, we must add it as a custom/advanced setting. not sure I / We can start testing all possible Postgres (9.X -> 13.X) or cockroach/yugabytedb/etc... and not MySQL ones... so this is a special config for a DBA/Optimizer, as SW can continue to work without it.

**Comment by kmlebedev** [comment]
> @kmlebedev fully backward compatible and without breaking new deployment, we must add it as a custom/advanced setting. not sure I / We can start testing all possible Postgres (9.X -> 13.X) or cockroach/yugabytedb/etc... and not MySQL ones... so this is a special config for a DBA/Optimizer, as SW can continue to work without it.

I disagree, since we have a problem https://github.com/chrislusf/seaweedfs/issues/1921
that is solved by this optimization, but in fact, only DBA can still fix

**Comment by LazyDBA247-Anyvision** [comment]
@kmlebedev the syntax I added in scaffold.go, should work in most of the mysql/postgres deployment. so maybe the only change that is needed is to add to the exception/error in the insertEntry is to notify the user that a fix can be done by using upsert and a good/working example exists in the scaffold?

on vanilla  mysql/postgres (last years versions) that example work.

**Comment by chrislusf** [comment]
Or how about just give user the option to "enableUpsert", and turn on by default?

**Comment by LazyDBA247-Anyvision** [comment]
@chrislusf that sounds a good option, but auto-enabled sounds risky to me. 
maybe we can add a "test/verify" the upsert command on startup?
and if no exception is being made?  something like 2x insert/upsert to "fake" path /test  (like /buckets  /queues ).
that will verify the DB support the syntax, with an exception in the log to update the syntax if not working...

@kmlebedev  that's what you wanted, yes?

**Comment by chrislusf** [comment]
> maybe we can add a "test/verify" the upsert command on startup?

That is safe but not really necessary. Upsert is command and most mysql/postgres dialect implementations should already support them. With defaulting enableUpsert to true, filers will fail fast if filer store does not support it, and not cause any issue.

**Comment by Aytuar** [comment]
[https://postgrespro.ru/docs/postgresql/9.6/sql-insert?lang=en
](url)
`
Tip
It is often preferable to use unique index inference rather than naming a constraint directly using ON CONFLICT ON CONSTRAINT constraint_name. Inference will continue to work correctly when the underlying index is replaced by another more or less equivalent index in an overlapping way, for example when using CREATE UNIQUE INDEX ... CONCURRENTLY before dropping the index being replaced.
`

**Pull request merged**